File: tests/test_preprocessing.cpp
#include <chrono>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <iterator>
#include <stdexcept>
#include <string>
using namespace std;
namespace fs = filesystem;

const string header = "Release Year,Title,Origin/Ethnicity,Director,Cast,Genre,Wiki Page,Plot\n";
const string movie = "2000,Pelicula,Peru,Ana,Actor,Drama,https://example.org/a,Trama\n";
fs::path folder;
string program;
int passed = 0;

void check(bool condition, const string& message) {
    if (!condition) throw runtime_error(message);
}
string readFile(const fs::path& path) {
    ifstream file(path, ios::binary);
    check(bool(file), "No se pudo leer " + path.string());
    return string(istreambuf_iterator<char>(file), istreambuf_iterator<char>());
}
// Protege rutas con espacios o apostrofes al ejecutar en macOS/Linux.
string shellQuote(const string& value) {
    string result = "'";
    for (char c : value) result += c == '\'' ? "'\"'\"'" : string(1, c);
    return result + "'";
}
int execute(const string& destination = "clean.csv", const string& report = "report.json") {
    string command = shellQuote(program) + " " + shellQuote((folder / "input.csv").string())
        + " " + shellQuote((folder / destination).string())
        + " " + shellQuote((folder / report).string())
        + " > " + shellQuote((folder / "log.txt").string()) + " 2>&1";
    return system(command.c_str());
}
int run(const string& content) {
    for (const string name : {"clean.csv", "report.json"}) fs::remove(folder / name);
    ofstream input(folder / "input.csv", ios::binary);
    input << content;
    input.close();
    check(bool(input), "Error creando entrada de prueba");
    int result = execute();
    check(readFile(folder / "input.csv") == content, "El original fue modificado");
    return result;
}
void reportContains(const string& expected) {
    check(readFile(folder / "report.json").find(expected) != string::npos,
          "Falta en el reporte: " + expected);
}
void pass(const string& name) { cout << "OK: " << name << '\n'; ++passed; }

int main(int argc, char** argv) {
    if (argc != 2) { cerr << "Uso: test_preprocessing ruta_al_limpiador\n"; return 1; }
    program = fs::absolute(argv[1]).string();
    folder = fs::temp_directory_path() / ("stream-c-tests-" +
        to_string(chrono::steady_clock::now().time_since_epoch().count()));
    try {
        check(fs::create_directory(folder), "No se pudo crear carpeta temporal");
        // 1. Texto complicado: comas, comillas, Unicode y saltos internos.
        string complex = "2000,\"El \"\"barco\"\", fantasma\",Perú,Ana,Actor,Drama,url,\"Primera línea\r\nSegunda línea\n日本語, \"\"sí\"\"\"";
        check(run(string("\xEF\xBB\xBF") + header + complex) == 0, "Fallo con texto complejo");
        check(readFile(folder / "clean.csv") == header + complex + "\n", "Se altero el texto");
        pass("comillas, multilinea, Unicode, BOM y final sin salto");

        // 2. Compara toda la salida con el resultado esperado, no solo contadores.
        string empty = "2000,Pelicula,Peru,\xC2\xA0, \t\xE2\x80\x83,,url,Trama\n";
        string fallback = "2000,,Peru,Ana,Actor,Drama,url,Trama\n";
        string extra = "2000,Otra,Peru,Ana,Actor,Drama,url,Trama,extra,extra2\n";
        string remake = "2020,Pelicula,Peru,Ana,Actor,Drama,https://example.org/a,Trama\n";
        check(run(header + movie + movie + empty + "2000,,Peru,Ana,Actor,Drama,,Trama\n"
                  + "2000,incompleta\n" + fallback + extra + remake) == 0, "Fallo en reglas");
        string expected = header + movie + "2000,Pelicula,Peru,unknown,unknown,unknown,url,Trama\n"
            + "2000,unknown,Peru,Ana,Actor,Drama,url,Trama\n"
            + "2000,Otra,Peru,Ana,Actor,Drama,url,Trama\n" + remake;
        check(readFile(folder / "clean.csv") == expected, "Limpieza incorrecta");
        for (const string field : {"\"registros_leidos\": 8", "\"registros_conservados\": 5",
             "\"registros_descartados\": 3", "\"duplicados\": 1", "\"columnas_extra_omitidas\": 2",
             "\"Cast\": 1", "\"registro\": 2,", "\"registro\": 4,", "\"registro\": 5,"}) reportContains(field);
        pass("vacios, duplicados, descartes, extras y titulos repetidos");

        // 3. Codificacion invalida.
        check(run(header + "2000,Pelicula,Peru,Ana," + string(1, char(0xFF)) + ",Drama,url,Trama\n") == 0,
              "Fallo al descartar UTF-8 invalido");
        reportContains("utf8_invalido_o_nul");
        check(readFile(folder / "clean.csv") == header, "Se conservo UTF-8 invalido");
        pass("UTF-8 invalido registrado como descarte");

        // 4. Nunca publicar una salida parcial si las comillas estan rotas.
        for (const string tail : {"2000,\"unclosed", "2000,ba\"d,x", "2000,\"closed\"oops,x"}) {
            check(run(header + movie + tail) != 0, "Se aceptaron comillas rotas");
            for (const string name : {"clean.csv", "report.json", "clean.csv.tmp", "report.json.tmp"})
                check(!fs::exists(folder / name), "Quedo un archivo parcial");
        }
        pass("comillas rotas sin archivos parciales");

        // 5. Cabecera incompatible.
        check(run("Title,Plot\na,b\n") != 0, "Se acepto una cabecera incorrecta");
        check(!fs::exists(folder / "clean.csv"), "Se genero salida con cabecera incorrecta");
        pass("cabecera incorrecta rechazada");

        // 6. Proteccion del original y de los resultados existentes.
        check(run(header + movie) == 0, "Fallo creando resultado");
        string previous = readFile(folder / "clean.csv");
        check(execute("clean.csv", "new-report.json") != 0, "Se sobrescribio el resultado");
        check(execute("input.csv", "new-report.json") != 0, "Se sobrescribio el original");
        check(readFile(folder / "clean.csv") == previous, "Cambio el resultado anterior");
        check(readFile(folder / "input.csv") == header + movie, "Cambio el original");
        pass("proteccion de originales y resultados existentes");

        // 7. Un archivo con cabecera y sin peliculas es valido.
        check(run(header) == 0, "Fallo con dataset sin peliculas");
        reportContains("\"registros_leidos\": 0");
        check(readFile(folder / "clean.csv") == header, "Salida vacia incorrecta");
        pass("dataset sin registros");
        fs::remove_all(folder);
        cout << passed << " pruebas aprobadas.\n";
    } catch (const exception& error) {
        cerr << "FALLO: " << error.what() << "\nArchivos para revisar: " << folder << '\n';
        return 1;
    }
}

File: tests/test_preprocessing.py
"""Pruebas de caja negra. El procesamiento productivo es exclusivamente C++."""
import csv
import io
import json
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest

BIN = str(Path(sys.argv.pop(1)).resolve())
HEADER = ['Release Year', 'Title', 'Origin/Ethnicity', 'Director',
          'Cast', 'Genre', 'Wiki Page', 'Plot']
BASE = ['2000', 'Película', 'Perú', 'Ana', 'Actor', 'Drama', 'https://example.org/a', 'Trama']


def encode(rows):
    s = io.StringIO(newline='')
    csv.writer(s).writerows([HEADER] + rows)
    return s.getvalue().encode('utf-8')


class PreprocessingTests(unittest.TestCase):
    def run_clean(self, content):
        folder = tempfile.TemporaryDirectory()
        self.addCleanup(folder.cleanup)
        root = Path(folder.name)
        source, output, report = (root / n for n in ['source.csv', 'clean.csv', 'report.json'])
        source.write_bytes(content)
        result = subprocess.run([BIN, str(source), str(output), str(report)], capture_output=True)
        self.assertEqual(source.read_bytes(), content, 'El original debe conservarse')
        return result, source, output, report

    def test_round_trip_multiline_quotes_unicode_and_bom(self):
        row = BASE.copy()
        row[1] = 'El "barco", fantasma'
        row[7] = 'Primera línea\r\nSegunda línea\n日本語, "sí"'
        result, _, out, report = self.run_clean(b'\xef\xbb\xbf' + encode([row]).rstrip(b'\r\n'))
        self.assertEqual(result.returncode, 0, result.stderr)
        with out.open(encoding='utf-8', newline='') as f:
            self.assertEqual(list(csv.reader(f)), [HEADER, row])
        self.assertEqual(json.loads(report.read_text())['registros_conservados'], 1)

    def test_rules_and_auditable_counts(self):
        empty = BASE.copy(); empty[3:6] = ['\u00a0', ' \t\u2003', '']
        missing = BASE.copy(); missing[1] = missing[6] = ''
        fallback = BASE.copy(); fallback[1] = ''
        extra = BASE.copy(); extra[1] = 'Otra'; extra += ['extra', 'extra2']
        remake = BASE.copy(); remake[0] = '2020'
        rows = [BASE, BASE, empty, missing, ['2000', 'incompleta'], fallback, extra, remake]
        result, _, out, report = self.run_clean(encode(rows))
        self.assertEqual(result.returncode, 0, result.stderr)
        stats = json.loads(report.read_text())
        self.assertEqual((stats['registros_leidos'], stats['registros_conservados'],
                          stats['registros_descartados'], stats['duplicados']), (8, 5, 3, 1))
        self.assertEqual(stats['columnas_extra_omitidas'], 2)
        self.assertEqual(stats['vacios_reemplazados_en_salida']['Cast'], 1)
        self.assertEqual([d['registro'] for d in stats['descartes']], [2, 4, 5])
        with out.open(newline='') as f:
            cleaned = list(csv.reader(f))[1:]
        self.assertTrue(all(len(r) == 8 for r in cleaned))
        self.assertEqual(cleaned[1][3:6], ['unknown'] * 3)
        self.assertEqual(cleaned[2][1], 'unknown')
        self.assertEqual(cleaned[-1][0], '2020')

    def test_invalid_utf8_is_reported(self):
        result, _, _, report = self.run_clean(encode([BASE]).replace(b'Actor', b'\xff'))
        self.assertEqual(result.returncode, 0, result.stderr)
        self.assertEqual(json.loads(report.read_text())['descartes'][0]['motivo'], 'utf8_invalido_o_nul')

    def test_broken_quotes_abort_without_partial_output(self):
        for tail in [b'2000,"unclosed', b'2000,ba"d,x', b'2000,"closed"oops,x']:
            with self.subTest(tail=tail):
                result, _, out, report = self.run_clean(encode([BASE]) + tail)
                self.assertNotEqual(result.returncode, 0)
                self.assertFalse(out.exists())
                self.assertFalse(report.exists())
                self.assertFalse(Path(str(out) + '.tmp').exists())

    def test_wrong_header_aborts(self):
        result, _, out, _ = self.run_clean(b'Title,Plot\na,b\n')
        self.assertNotEqual(result.returncode, 0)
        self.assertFalse(out.exists())

    def test_existing_output_and_original_are_protected(self):
        result, source, out, report = self.run_clean(encode([BASE]))
        self.assertEqual(result.returncode, 0)
        before = out.read_bytes()
        for dest in [out, source]:
            again = subprocess.run([BIN, str(source), str(dest), str(report)], capture_output=True)
            self.assertNotEqual(again.returncode, 0)
        self.assertEqual(out.read_bytes(), before)
        self.assertEqual(source.read_bytes(), encode([BASE]))

    def test_header_only(self):
        result, _, _, report = self.run_clean(encode([]))
        self.assertEqual(result.returncode, 0)
        self.assertEqual(json.loads(report.read_text())['registros_leidos'], 0)


if __name__ == '__main__':
    unittest.main()
